src/models/StyledComponent.js

```diff
@@ -64,14 +64,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
         return context
       }
 
-      this.attrs = Object.keys(attrs).reduce((acc, key) => {
+      this.attrs = this.resolveComponentAttrs(attrs, context)
+
+      return { ...context, ...this.attrs }
+    }
+
+    resolveComponentAttrs(attrs: any, context: any) {
+      if (typeof attrs === 'function') {
+        return attrs(context)
+      }
+
+      return Object.keys(attrs).reduce((acc, key) => {
```

Comment:
I'm wondering... We could abstract this part of the `attrs` resolution to `constructWithOptions` now. This would help fixing the above described edge case and would result in always being able to run `attrs(context)` here, which helps v8 in optimising it.